pdns/syncres.cc

```diff
@@ -2117,11 +2255,22 @@ RCode::rcodes_ SyncRes::updateCacheFromRecords(unsigned int depth, LWResult& lwr
             recordState = validateDNSKeys(i->first.name, i->second.records, i->second.signatures, depth);
           }
           else {
-            LOG(d_prefix<<"Validating non-additional record for "<<i->first.name<<endl);
-            recordState = validateRecordsWithSigs(depth, qname, qtype, i->first.name, i->second.records, i->second.signatures);
-            /* we might have missed a cut (zone cut within the same auth servers), causing the NS query for an Insecure zone to seem Bogus during zone cut determination */
-            if (qtype == QType::NS && i->second.signatures.empty() && recordState == Bogus && haveExactValidationStatus(i->first.name) && getValidationStatus(i->first.name) == Indeterminate) {
-              recordState = Indeterminate;
+            /*
+             * RFC 6672 section 5.3.1
+             *  In any response, a signed DNAME RR indicates a non-terminal
+             *  redirection of the query.  There might or might not be a server-
+             *  synthesized CNAME in the answer section; if there is, the CNAME will
+             *  never be signed.  For a DNSSEC validator, verification of the DNAME
+             *  RR and then that the CNAME was properly synthesized is sufficient
+             *  proof.
+             */
```

Comment:
So we don't really validate that the `CNAME` was properly synthesized below, we fix it if it wasn't. Is that the correct behavior?

Reply:
RFC 6672 section 3.4.1 "Resolver Algorithm", in step 4 "Analyze the response"

>        D.  If the response shows a DNAME and that is not the answer
           itself, cache the DNAME (upon successful DNSSEC validation if
           the client is a validating resolver).  If substitution of the
           DNAME's target name for its owner name in the SNAME would
           overflow the legal size for a domain name, return an
           implementation-dependent error to the application; otherwise,
           perform the substitution and go to step 1.

Reply:
Ans it looks like we expect the CNAME to be in the response, I'll fix that as well